File: pkg/dev_compiler/test/codegen_expected/lib/html/typed_arrays_simd_test.js
dart_library.library('lib/html/typed_arrays_simd_test', null, /* Imports */[
  'dart_sdk',
  'unittest'
], function load__typed_arrays_simd_test(exports, dart_sdk, unittest) {
  'use strict';
  const core = dart_sdk.core;
  const html = dart_sdk.html;
  const typed_data = dart_sdk.typed_data;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const src__matcher__numeric_matchers = unittest.src__matcher__numeric_matchers;
  const html_config = unittest.html_config;
  const unittest$ = unittest.unittest;
  const src__matcher__expect = unittest.src__matcher__expect;
  const src__matcher__core_matchers = unittest.src__matcher__core_matchers;
  const typed_arrays_simd_test = Object.create(null);
  let dynamicTodynamic = () => (dynamicTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [dart.dynamic])))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  typed_arrays_simd_test._FLOATING_POINT_ERROR = 1e-10;
  typed_arrays_simd_test.floatEquals = function(value) {
    return src__matcher__numeric_matchers.closeTo(core.num._check(value), typed_arrays_simd_test._FLOATING_POINT_ERROR);
  };
  dart.fn(typed_arrays_simd_test.floatEquals, dynamicTodynamic());
  typed_arrays_simd_test.MyFloat32x4 = class MyFloat32x4 extends core.Object {
    new() {
      this.x = 0.0;
      this.y = 0.0;
      this.z = 0.0;
      this.w = 0.0;
    }
  };
  typed_arrays_simd_test.main = function() {
    html_config.useHtmlConfiguration();
    if (!dart.test(html.Platform.supportsTypedData)) {
      return;
    }
    unittest$.test('test Float32x4', dart.fn(() => {
      if (dart.test(html.Platform.supportsSimd)) {
        let val = typed_data.Float32x4.new(1.0, 2.0, 3.0, 4.0);
        src__matcher__expect.expect(val.x, typed_arrays_simd_test.floatEquals(1.0));
        src__matcher__expect.expect(val.y, typed_arrays_simd_test.floatEquals(2.0));
        src__matcher__expect.expect(val.z, typed_arrays_simd_test.floatEquals(3.0));
        src__matcher__expect.expect(val.w, typed_arrays_simd_test.floatEquals(4.0));
        let val2 = val['+'](val);
        src__matcher__expect.expect(val2.x, typed_arrays_simd_test.floatEquals(2.0));
        src__matcher__expect.expect(val2.y, typed_arrays_simd_test.floatEquals(4.0));
        src__matcher__expect.expect(val2.z, typed_arrays_simd_test.floatEquals(6.0));
        src__matcher__expect.expect(val2.w, typed_arrays_simd_test.floatEquals(8.0));
      }
    }, VoidTodynamic()));
    unittest$.test('test Float32x4List', dart.fn(() => {
      let counter = null;
      let list = typed_data.Float32List.new(12);
      for (let i = 0; i < dart.notNull(list[dartx.length]); ++i) {
        list[dartx.set](i, i * 1.0);
      }
      if (dart.test(html.Platform.supportsSimd)) {
        counter = typed_data.Float32x4.zero();
        let simdlist = typed_data.Float32x4List.view(list[dartx.buffer]);
        for (let i = 0; i < dart.notNull(simdlist.length); ++i) {
          counter = dart.dsend(counter, '+', simdlist.get(i));
        }
      } else {
        counter = new typed_arrays_simd_test.MyFloat32x4();
        for (let i = 0; i < dart.notNull(list[dartx.length]); i = i + 4) {
          dart.dput(counter, 'x', dart.dsend(dart.dload(counter, 'x'), '+', list[dartx.get](i)));
          dart.dput(counter, 'y', dart.dsend(dart.dload(counter, 'y'), '+', list[dartx.get](i + 1)));
          dart.dput(counter, 'z', dart.dsend(dart.dload(counter, 'z'), '+', list[dartx.get](i + 2)));
          dart.dput(counter, 'w', dart.dsend(dart.dload(counter, 'w'), '+', list[dartx.get](i + 3)));
        }
      }
      src__matcher__expect.expect(dart.dload(counter, 'x'), typed_arrays_simd_test.floatEquals(12.0));
      src__matcher__expect.expect(dart.dload(counter, 'y'), typed_arrays_simd_test.floatEquals(15.0));
      src__matcher__expect.expect(dart.dload(counter, 'z'), typed_arrays_simd_test.floatEquals(18.0));
      src__matcher__expect.expect(dart.dload(counter, 'w'), typed_arrays_simd_test.floatEquals(21.0));
    }, VoidTodynamic()));
    unittest$.test('test Int32x4', dart.fn(() => {
      if (dart.test(html.Platform.supportsSimd)) {
        let val = typed_data.Int32x4.new(1, 2, 3, 4);
        src__matcher__expect.expect(val.x, src__matcher__core_matchers.equals(1));
        src__matcher__expect.expect(val.y, src__matcher__core_matchers.equals(2));
        src__matcher__expect.expect(val.z, src__matcher__core_matchers.equals(3));
        src__matcher__expect.expect(val.w, src__matcher__core_matchers.equals(4));
        let val2 = val['^'](val);
        src__matcher__expect.expect(val2.x, src__matcher__core_matchers.equals(0));
        src__matcher__expect.expect(val2.y, src__matcher__core_matchers.equals(0));
        src__matcher__expect.expect(val2.z, src__matcher__core_matchers.equals(0));
        src__matcher__expect.expect(val2.w, src__matcher__core_matchers.equals(0));
      }
    }, VoidTodynamic()));
  };
  dart.fn(typed_arrays_simd_test.main, VoidTodynamic());
  // Exports:
  exports.typed_arrays_simd_test = typed_arrays_simd_test;
});

File: tests/html/typed_arrays_simd_test.dart
// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

library TypedArraysSimdTest;

import 'package:unittest/unittest.dart';
import 'package:unittest/html_config.dart';
import 'dart:html';
import 'dart:typed_data';

const _FLOATING_POINT_ERROR = 0.0000000001;
floatEquals(value) => closeTo(value, _FLOATING_POINT_ERROR);

class MyFloat32x4 {
  num x = 0.0;
  num y = 0.0;
  num z = 0.0;
  num w = 0.0;
}

main() {
  useHtmlConfiguration();

  // Only perform tests if ArrayBuffer is supported.
  if (!Platform.supportsTypedData) {
    return;
  }

  test('test Float32x4', () {
    if (Platform.supportsSimd) {
      final val = new Float32x4(1.0, 2.0, 3.0, 4.0);
      expect(val.x, floatEquals(1.0));
      expect(val.y, floatEquals(2.0));
      expect(val.z, floatEquals(3.0));
      expect(val.w, floatEquals(4.0));
      final val2 = val + val;
      expect(val2.x, floatEquals(2.0));
      expect(val2.y, floatEquals(4.0));
      expect(val2.z, floatEquals(6.0));
      expect(val2.w, floatEquals(8.0));
    }
  });

  test('test Float32x4List', () {
    var counter;
    final list = new Float32List(12);
    for (int i = 0; i < list.length; ++i) {
      list[i] = i * 1.0;
    }
    if (Platform.supportsSimd) {
      counter = new Float32x4.zero();
      final simdlist = new Float32x4List.view(list.buffer);
      for (int i = 0; i < simdlist.length; ++i) {
        counter += simdlist[i];
      }
    } else {
      counter = new MyFloat32x4();
      for (int i = 0; i < list.length; i += 4) {
        counter.x += list[i];
        counter.y += list[i + 1];
        counter.z += list[i + 2];
        counter.w += list[i + 3];
      }
    }
    expect(counter.x, floatEquals(12.0));
    expect(counter.y, floatEquals(15.0));
    expect(counter.z, floatEquals(18.0));
    expect(counter.w, floatEquals(21.0));
  });

  test('test Int32x4', () {
    if (Platform.supportsSimd) {
      final val = new Int32x4(1, 2, 3, 4);
      expect(val.x, equals(1));
      expect(val.y, equals(2));
      expect(val.z, equals(3));
      expect(val.w, equals(4));
      final val2 = val ^ val;
      expect(val2.x, equals(0));
      expect(val2.y, equals(0));
      expect(val2.z, equals(0));
      expect(val2.w, equals(0));
    }
  });
}
